crates/core/src/dht/dht_store.rs

```diff
@@ -279,30 +290,69 @@ impl DhtStore {
                 }
                 false
             })
-            .map(|(pending, maybe_delay)| {
-                (
-                    pending,
-                    maybe_delay
-                        .map(|(_time, duration)| Some(duration))
-                        .unwrap_or(None),
-                )
+            .map(|PendingValidationWithTimeout { pending, timeout }| {
+                (pending, timeout.map(|t| Some(t.delay)).unwrap_or(None))
             })
             .next()
     }
 
     pub(crate) fn has_queued_holding_workflow(&self, pending: &PendingValidation) -> bool {
-        self.queued_holding_workflows
-            .iter()
-            .any(|(current, _)| current == pending)
+        self.queued_holding_workflows.iter().any(
+            |PendingValidationWithTimeout {
+                 pending: current, ..
+             }| current == pending,
+        )
     }
 
-    pub(crate) fn queued_holding_workflows(
-        &self,
-    ) -> &VecDeque<(PendingValidation, Option<(SystemTime, Duration)>)> {
+    pub(crate) fn queued_holding_workflows(&self) -> &VecDeque<PendingValidationWithTimeout> {
         &self.queued_holding_workflows
     }
 }
 
+use petgraph::{algo::is_cyclic_directed, graph::DiGraph, prelude::NodeIndex, Direction::Outgoing};
+use std::collections::HashMap;
+
+fn get_free_dependencies<I>(pending: &I) -> Vec<PendingValidationWithTimeout>
+where
+    I: IntoIterator<Item = PendingValidationWithTimeout> + Clone,
+{
+    let mut graph = DiGraph::<(), ()>::new();
+    let mut index_map: HashMap<Address, NodeIndex> = HashMap::new();
+    let mut index_reverse_map: HashMap<NodeIndex, PendingValidationWithTimeout> = HashMap::new();
+
+    // add the nodes
+    for p in pending.clone() {
+        let node_index = graph.add_node(());
+        index_map.insert(p.pending.entry_with_header.entry.address(), node_index);
+        index_reverse_map.insert(node_index, p);
+    }
+
+    // add the edges
+    for p in pending.clone() {
+        let from = index_map
+            .get(&p.pending.entry_with_header.entry.address())
+            .expect("we literally just added this");
+        for to_addr in p.pending.dependencies.clone() {
+            // only add the dependencies that are also in the pending validation list
+            if let Some(to) = index_map.get(&to_addr) {
+                graph.add_edge(*from, *to, ());
+            }
+        }
+    }
+
+    if is_cyclic_directed(&graph) {
+        // this might be expensive..
+        panic!("Cyclic validation dependencies detected!!")
```

Comment:
I'm really not sure what we should do here. If there are cyclic validation dependencies the DNA is cooked. I don't think there is any way to recover from this.

Reply:
This isn't about the DNA it's about the data.  We can't panic because that means if someone malicious finds a way to send people fake headers and entries that create a cycle it kills their conductor.   It think what we have to do here is declare those entries invalid and reject them.  This then become part of validation.   I'm ok with adding this as follow-up, but it's pretty high priority to get right before it fades in our memory.

Reply:
Agree with @zippy. I would actually want to have this panic be an error already before merging.

Reply:
Hm, ok, I see that it is not easy to deal with this here. So I, too, don't want to block merging this, but we need to handle this somehow. Panicking will might kill the whole conductor, also other DNAs/hApps that would work. Maybe we can detect the cycle and at least continue validating the other entries and if there are no other entries just trying to validate any of the entries in the cycle ?